src/osd/PeeringState.cc

```diff
@@ -2118,8 +2118,8 @@ void PeeringState::calc_replicated_acting_stretch(
    */
   while (!aheap.is_empty() && want->size() < pool.info.size) {
     auto next = aheap.pop();
-    pop_ancestor(next.get());
     if (next.get().get_num_selected() < bucket_max) {
+      pop_ancestor(next.get());
```

Comment:
Ouch, good catch. I'm trying to think through the implications of this bug. As you state on the ticket, we won't break our consistency rules — does this just lead to too many OSDs in an acting set? Will we be unable to go active because we don't put enough in from other buckets? Or do we end up with 3 in one DC and 1 in the other?

Or do we not actually hit this in our current deployments, because we are only using two buckets and "our heap sorting always pulls from ancestors with the least number of included OSDs" (in which case, does it have any impact ever?).

Reply:
The bug leads to too many OSDs from the same bucket in the acting set, which is what we want to avoid.
Whether we can go active depends on the minimum size we set our pool to.

I think we never really felt the impact of this bug because of the low probability of a scenario where we backfill to a third OSD in DC1, accepts writes, fail DC1 and DC2 comes back up simultaneously/

Reply:
If we can actually hit it with two sites, I think it needs to go in a separate commit and be backported to relevant releases?

Reply:
I think that's a good idea and sensible. Will move it out

Reply:
dropped the commit and moved it to https://github.com/ceph/ceph/pull/57381